src/common/serialization.h

```diff
@@ -0,0 +1,319 @@
+ * \file serialization.h
+ * \brief Serialization of some STL and nnvm data-structures
+ * \author Clement Fuji Tsang
+ */
+
+#ifndef MXNET_COMMON_SERIALIZATION_H_
+#define MXNET_COMMON_SERIALIZATION_H_
+
+#include <dmlc/logging.h>
+#include <mxnet/graph_attr_types.h>
+#include <nnvm/graph_attr_types.h>
+#include <nnvm/tuple.h>
+
+#include <cstring>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+
+namespace mxnet {
+namespace common {
+
+template<typename T>
+inline size_t serialized_size(const T& obj);
+
+template<typename T>
+inline size_t serialized_size(const nnvm::Tuple<T>& obj);
+
+template<typename K, typename V>
+inline size_t serialized_size(const std::map<K, V>& obj);
+
+template<>
+inline size_t serialized_size(const std::string& obj);
+
+template<typename... Args>
+inline size_t serialized_size(const std::tuple<Args...>& obj);
+
+template<typename T>
+inline void serialize(const T& obj, char** buffer);
+
+template<typename T>
+inline void serialize(const nnvm::Tuple<T>& obj, char** buffer);
+
+template<typename K, typename V>
+inline void serialize(const std::map<K, V>& obj, char** buffer);
+
+template<>
+inline void serialize(const std::string& obj, char** buffer);
+
+template<typename... Args>
+inline void serialize(const std::tuple<Args...>& obj, char** buffer);
+
+template<typename T>
+inline void deserialize(T* obj, const std::string& buffer, size_t* curr_pos);
+
+template<typename T>
+inline void deserialize(nnvm::Tuple<T>* obj, const std::string& buffer, size_t* curr_pos);
+
+template<typename K, typename V>
+inline void deserialize(std::map<K, V>* obj, const std::string& buffer, size_t* curr_pos);
+
+template<>
+inline void deserialize(std::string* obj, const std::string& buffer, size_t* curr_pos);
+
+template<typename... Args>
+inline void deserialize(std::tuple<Args...>* obj, const std::string& buffer, size_t* curr_pos);
+
+
+template<typename T>
+struct is_container {
+  static const bool value = !std::is_pod<T>::value;
+};
+
+template<typename T>
+inline size_t serialized_size(const T& obj) {
+  return sizeof(T);
+}
+
+template<typename T>
+inline size_t serialized_size(const nnvm::Tuple<T>& obj) {
+  if (is_container<T>::value) {
+    size_t sum_val = 4;
+    for (auto& el : obj) {
+      sum_val += serialized_size(el);
+    }
+    return sum_val;
+  } else {
+    return 4 + (obj.ndim() * sizeof(T));
+  }
+}
+
+template<typename K, typename V>
+inline size_t serialized_size(const std::map<K, V>& obj) {
+  size_t sum_val = 4;
+  if (is_container<K>::value && is_container<V>::value) {
+    for (auto p : obj) {
+      sum_val += serialized_size(p.first) + serialized_size(p.second);
+    }
+  } else if (is_container<K>::value) {
+    for (auto p : obj) {
+      sum_val += serialized_size(p.first);
+    }
+    sum_val += sizeof(V) * obj.size();
+  } else if (is_container<K>::value) {
```

Comment:
Should we be checking is_container<K> twice here?  I think one of these branches might intend to check V instead of K.